lib/styles.js

```diff
@@ -340,4 +342,145 @@ module.exports = function (expect) {
+                lastPart = null;
+            } else if (lastPart) {
+                diffLines.push(lastPart);
+                lastPart = part;
+            } else {
+                lastPart = part;
+            }
+        });
+        if (lastPart) {
+            diffLines.push(lastPart);
+        }
+
+        diffLines.forEach(function (part, index) {
+            var endsWithNewline = /\n$/.test(part.value);
+            var value;
+            if (part.replaced) {
+                var oldLine = this.clone();
+                var newLine = this.clone();
+                var oldValue = part.oldValue;
+                var newValue = part.newValue;
+                var oldEndsWithNewline = oldValue.slice(-1) === '\n';
+                var newEndsWithNewline = newValue.slice(-1) === '\n';
+                if (oldEndsWithNewline) {
+                    oldValue = oldValue.slice(0, -1);
+                }
+                if (newEndsWithNewline) {
+                    newValue = newValue.slice(0, -1);
+                }
+
+                stringDiff['diff' + type](oldValue, newValue).forEach(function (part) {
+                    if (part.added) {
+                        newLine.diffFragment(part.value, 'diffAddedHighlight', 'diffAddedSpecialChar', options.markUpSpecialCharacters);
+                    } else if (part.removed) {
+                        oldLine.diffFragment(part.value, 'diffRemovedHighlight', 'diffRemovedSpecialChar', options.markUpSpecialCharacters);
+                    } else {
+                        newLine.diffAddedLine(part.value);
+                        oldLine.diffRemovedLine(part.value);
+                    }
+                });
+                this.alt({
+                    text: function () {
+                        oldLine.prependLinesWith(this.clone().diffRemovedLine('-'));
+                        newLine.prependLinesWith(this.clone().diffAddedLine('+'));
+                    },
+                    fallback: function () {}
+                });
+
+                if (oldEndsWithNewline && !newEndsWithNewline) {
+                    oldLine.diffRemovedSpecialChar('\\n');
+                }
+
+                if (newEndsWithNewline && !oldEndsWithNewline) {
+                    newLine.diffAddedSpecialChar('\\n');
+                }
+
+                this.append(oldLine).nl().append(newLine);
+                if (oldEndsWithNewline && index < diffLines.length - 1) {
+                    this.nl();
+                }
+            } else if (part.added) {
+                value = endsWithNewline ?
+                    part.value.slice(0, -1) :
+                    part.value;
+
+                this.append(function () {
+                    this.diffFragment(value, 'diffAddedLine', undefined, options.markUpSpecialCharacters);
+                    var outer = this;
+                    this.alt({
+                        text: function () {
+                            outer.prependLinesWith(function () {
```

Comment:
I think that would be achievable. I want to rewrite most of it, as there are a lot of code smells, especially after I introduced those `alt` constructs. MagicPen has improved a lot since `utils.diffStrings` was originally written.

The problem with `alt` delegating to `append` that I described above wouldn't be a problem if we could get rid of the `prependLinesWith`. Do you consider that an antipattern in general?